### ISSUE TITLE
Mapping of numbers and non nullable booleans with an Oracle DB

### ISSUE THREAD

**Issue by SME82** [ErikEJ/EFCorePowerTools#1443]
I used a "database first" approach with an Oracle database. I had a couple of problems with the models generated.

All non-nullable booleans where generated as bool? properties. I found the option "Remove SQL default from bool columns" is supposed to change this behaviour. But for me it does nothing at all. I get the same code generated with and without the option.
I tested this with Columns that are default true and columns that are default false:

**"AUTOMATIC_REPLY" NUMBER(1,0) DEFAULT 1 NOT NULL ENABLE,**

=> generated Property: **public bool? AutomaticReply { get; set; }**

**"CARRIER" NUMBER(1,0) DEFAULT 0 NOT NULL ENABLE,**

=> generated Property: **public bool? Carrier { get; set; }**

Another problem are the properties generated for small numbers. For example:

**"CLASS_ID" NUMBER(3,0),**

=> generated Property: **public byte ClassId { get; set; }**

This mapping is correct for a code-first approach. A byte is indeed a Number(3,0).
But it does not work the other way around. A Number(3,0) is not always a byte. Number(3,0) ranges from -999 to 999 whereas byte ranges from 0 to 255.

The same problem applies to Number(5,0). Number(5,0) is mapped to short. Again correct for code-first but wrong for database-first.
short => values –32.768 to 32.767
number(5,0) => values -99.999 to 99.999

**Technical details:**

- .NetCore 6
- Oracle database(Oracle 19c)
- Microsoft.EntityFrameworkCore 6.0.5
- EF Core Power Tools version: 2.5.1007
- Visual Studio 2022)

**Comment by ErikEJ** [comment]
The first issue could possibly be fixed by adding support for Oracle in the code that removes the default sql.

For the three remaining issues this appears to be caused by the Oracle provider. Maybe @alexkeh can tell you where to report it.

**Comment by alexkeh** [comment]
@SME82 The Oracle team is aware of this issue. Currently, there is no way to create one set of data type mappings for code first and another for DB first. EF Core only allows one mapping to be used. Thus, ODP.NET is in the unenviable position of choosing which data type mappings should best match. Since code first is more common, the default mapping is better suited for code first.

If EF Core itself allows for provider writers to create two sets of data type mappings in the future, Oracle will be sure to use that capability. I know other developers have also requested a better default mapping for DB first.

**Comment by SME82** [comment]
-----------------------------------------------------------
Attention: Non-Delivery Report
-----------------------------------------------------------

This report is generated by the email server at:

       kv-rlp.de

The message with subject:

       "Re: [ErikEJ/EFCorePowerTools] Mapping of numbers and non nullable booleans with an Oracle DB (Issue #1443)"

and attached to this report was not delivered to 
the following recipients:

Address: ***@***.***
Reason:  550-5.7.1 This email was rejected because it violates our security policy (550)
--------------

**Comment by roji** [comment]
> Currently, there is no way to create one set of data type mappings for code first and another for DB first. EF Core only allows one mapping to be used. Thus, ODP.NET is in the unenviable position of choosing which data type mappings should best match.
> [...]
> If EF Core itself allows for provider writers to create two sets of data type mappings in the future, Oracle will be sure to use that capability.

I'm not quite sure what this means... EF's RelationalTypeMappingSource supports resolving type mappings either given a CLR type (used in "code first"), or given a string database type (this is used in scaffolding) (that's a very rough high-level description, but...). These are two separate lookup paths within the type mapping source, and there's no requirement that type mappings roundtrip (i.e. that given CLR type X that maps to database type Y, Y has to also map back to X).

In other words, it should be perfectly possible to map CLR byte to Oracle `Number(3,0)` (in "code first"), but when a `Number(3,0)` column is scaffolded, to get a different type (e.g. short).

Am I missing something, or is there some problem in EF I haven't understood?

**Comment by alexkeh** [comment]
@roji It is possible to change the data type mapping through post-code generation customization. 

I don't believe it's possible for a provider writer to create a provider that uses one default convention for DB first and a different default mapping convention for code first. That's what my dev team has relayed to me in the past since other customers have asked for the same change to ODP.NET EF Core's default DB first mapping.

Is this now possible?

**Comment by roji** [comment]
@alexkeh can you describe exactly what it is you're looking to do, with examples?

As I wrote above, it's perfectly doable to map CLR type X to some database Y (code-first), but when a column of database Y is scaffolded, to map it to a different CLR type Z. If you provide more context, I'd be able to understand better and possibly help.

**Comment by ErikEJ** [comment]
@alexkeh as @roji describes it is possible to have two napping resolutions in your type mapper. But maybe your devs were not aware of this. 

Now if only the Oracle provider was open source, then the community could help 😁

**Comment by SME82** [comment]
Seems our system security blocked the github notifications. Whoops...
I just wanted to thank you all for the quick replies.

**Comment by alexkeh** [comment]
Thanks for the info @roji. I'm following up with my dev team and will be in touch.

**Comment by ErikEJ** [comment]
@SME82 If you know of a way for me to easily spin up an Oracle instance and can provide a full CREATE TABLE script I can attempt a fix. Otherwise a PR would be appreciated.

**Comment by SME82** [comment]
@ErikEJ I cobbled this script together from a couple of tables. I removed all the foreign key constraints.
I am not sure what you mean by "PR". Problem report, Pull request? Unfortunately I will be unavailable the coming weeks (on vacation).

CREATE SEQUENCE  "SEQ_AP_OWNER"  MINVALUE 1 MAXVALUE 9999999999999999999999999999 INCREMENT BY 1 START WITH 35004 CACHE 10 NOORDER  NOCYCLE  NOKEEP  NOSCALE  GLOBAL ;

CREATE TABLE "AP_OWNER" 
(	"ID" NUMBER NOT NULL ENABLE, 
	"PID" NUMBER,
	"ISFOOTER" NUMBER(1,0),
	"AUTOMATIC_REPLY" NUMBER(1,0) DEFAULT 1 NOT NULL ENABLE,
	"CARRIER" NUMBER(1,0) DEFAULT 0 NOT NULL ENABLE,
	"CLASS_ID" NUMBER(3,0),
	"PROCESS_STATE_ID" NUMBER(5,0),
	"CASE_NR" NUMBER(9,0) NOT NULL ENABLE,
	 CONSTRAINT "AP_OWNER_PK" PRIMARY KEY ("ID") ENABLE
);

CREATE OR REPLACE EDITIONABLE TRIGGER "AP_OWNER_ID" 
BEFORE INSERT ON AP_OWNER
REFERENCING NEW AS NEW
FOR EACH ROW
BEGIN
	SELECT SEQ_AP_OWNER.NEXTVAL
	INTO :NEW.ID
	FROM DUAL;
END;

/
ALTER TRIGGER "AP_OWNER_ID" ENABLE;

**Comment by ErikEJ** [comment]
PR means pull request. I will give it a try soon.

**Comment by ErikEJ** [comment]
@alexkeh I need a connection string only approach, and followed your [advice here](https://medium.com/oracledevs/securely-connect-net-to-oracle-autonomous-database-without-a-wallet-nor-client-configuration-files-a1438893240a) to get there

**Comment by ErikEJ** [comment]
I have enabled the "Remove SQL default from bool columns" option to the latest daily build. Please try it out and let me know the outcome.

**Comment by alexkeh** [comment]
@roji My dev team is reviewing this code area. We'll get back to you by end of this week. We may contact you via email instead of continuing this thread if it looks to be an involved discussion. This feature request isn't specific to EF Core Power Tools functionality.

**Comment by roji** [comment]
@alexkeh sounds good, let me know if you need any help.

**Comment by ErikEJ** [comment]
@alexkeh also happy to assist!

**Comment by ErikEJ** [comment]
@SME82 did you have a chance to test the latest daily build?

**Comment by ErikEJ** [comment]
@SME82 I am going to close this. Feel free to re-open if the bool default fix does not work for you

**Comment by SME82** [comment]
Sorry for the long silence. I was on vacation and the following weeks I had to work on a couple of issues that occured while I was gone.

I had the opportunity to test it now. The option "Remove SQL default from bool columns" is working as intended. I get the non nullable boolean properties I wanted.

The other issue remains. Number(3,0) is still mapped to byte. However this is less of a problem for me, since I can increase the column sizes as a workaround.

Thanks a lot for fixing the problem with the boolen property generation.

**Comment by dleggen** [comment]
@ErikEJ would it be possible if varchar(2) 1 byte to be converted to bools as well with values of Y(y) or N(n)? Where N is the default and would be false. That is what we are using in our Oracle database.

**Comment by ErikEJ** [comment]
I dont think that would be a generally viable solution. Sounds like it is something you should discuss with the Oracle provider maintainers.

**Comment by ErikEJ** [comment]
@dleggen and this issue is closed
